apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,702 @@
+        public override async Task<Instance> GetInstanceMetadataAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var response = await RequestMetadata("instance?recursive=true", cancellationToken).ConfigureAwait(false);
+            dynamic obj = JObject.Parse(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
+            try
+            {
+                return new Instance
+                {
+                    CpuPlatform = obj.cpuPlatform,
+                    Description = obj.description,
+                    Disks = ((JArray)obj.disks).ToList(disk => new AttachedDisk
+                    {
+                        DeviceName = disk.deviceName,
+                        Index = disk.index,
+                        Mode = disk.mode,
+                        Type = disk.type
+                    }),
+                    Id = obj.id,
+                    MachineType = obj.machineType,
+                    Metadata = new Apis.Compute.v1.Data.Metadata()
+                    {
+                        Items = ((JObject)obj.attributes).ToList(attribute => new ItemsData
+                        {
+                            Key = attribute.Name,
+                            Value = attribute.Value
+                        })
+                    },
+                    Name = obj.hostname,
+                    NetworkInterfaces = ((JArray)obj.networkInterfaces).ToList(networkInterface => new NetworkInterface
+                    {
+                        AccessConfigs = ((JArray)networkInterface.accessConfigs).ToList(accessConfig => new AccessConfig
+                        {
+                            NatIP = accessConfig.externalIp,
+                            Type = accessConfig.type
+                        }),
+                        Network = networkInterface.network,
+                        NetworkIP = networkInterface.ip
+                    }),
+                    Scheduling = new Scheduling
+                    {
+                        AutomaticRestart = obj.scheduling.automaticRestart,
+                        OnHostMaintenance = obj.scheduling.onHostMaintenance,
+                        Preemptible = obj.scheduling.preemptible
+                    },
+                    ServiceAccounts = ((JObject)obj.serviceAccounts).ToList(serviceAccount => new ServiceAccount
+                    {
+                        Email = serviceAccount.Name,
+                        Scopes = ((JArray)serviceAccount.Value.scopes).ToList(scope => (string)scope.Value)
+                    }),
+                    Tags = new Tags
+                    {
+                        Items = ((JArray)obj.tags).ToList(tag => (string)tag.Value)
+                    },
+                    Zone = obj.zone
+                };
+            }
+            catch (Exception e)
+            {
+                // TODO: Should we use a custom exception type for this?
```

Comment:
Yeah that was a bit lazy on my part. This actually doesn't even do what I was hoping for: to catch exceptions when expected keys are missing, and I just didn't test it. I didn't realize the exception would be caught internally and null would be returned for missing keys. I've changed it to catch RuntimeBinderException for now to get data type mismatches, but I've added TODOs to figure out a better solution for detecting missing keys.